### DISCUSSION TITLE
Are the included engines available during offline operations?

### DISCUSSION THREAD

**Discussion by rajibando** [KoljaB/RealtimeTTS#217] [Q&A] [answered]
Hello, Mr. Kolja Beigel,

Thank you for your nice program, RealtimeTTS.

Watched the demo: Short_RealtimeTTS_Demo.mov

It covered the following engines sequentially: Coqui TTS, Azure TTS, Elevenlabs and System TTS. 

The rest: gTTS and Parler TTS weren't included in the demo. But no issues. The four engines demonstrated are sufficient for my purpose.

The first three, Coqui TTS, Azure TTS and Elevenlabs are very good and natural.

But are they available during offline operations?

Also, could the text pasted on the textbox for reading as a steam be endless? Any maximum limit?

I would try:
`pip install realtimetts[azure,elevenlabs,coqui]` once I hear from members who have already used this wonderful program.

**Comment by KoljaB** [comment, marked as answer]
Hi,

> The rest: gTTS and Parler TTS weren't included in the demo. 

the demo is old, gTTS and Parler were integrated after that. I won't create another video for every new engine.

> The first three, Coqui TTS, Azure TTS and Elevenlabs are very good and natural. But are they available during offline operations?

Azure TTS and Elevenlabs are APIs and can't be made "offline" because they a services. Coqui works offline.

> Also, could the text pasted on the textbox for reading as a steam be endless? Any maximum limit?

RealtimeTTS does not impose a specific limit on text size. In practice, the constraints are only theoretical and should be enormous, dictated by your system's memory capacity or the capabilities of the Textbox UI component. For most use cases, these limits are unlikely to be an issue.

> I would try: pip install realtimetts[azure,elevenlabs,coqui] once I hear from members who have already used this wonderful program.

RealtimeTTS has [100k downloads](https://pepy.tech/projects/realtimetts) and is used by huge projects like [Open Interpreter 01](https://github.com/OpenInterpreter/01) or [Neuro](https://github.com/kimjammer/Neuro).

**Reply by rajibando** [reply]
Thank you, Mr. Beigel, for replying to my question. Please accept my admiration.

Are there any offline TTS engine that could be used other than `System TTS`? `System TTS` is not very human-like.

> RealtimeTTS does not impose a specific limit on text size ...

Great to hear!

>  RealtimeTTS has [100k downloads](https://pepy.tech/projects/realtimetts) and is used by huge projects ...

Wow! Great!

The flipside is that the installation size appears huge, The download has exceeded 1GB filesize. Of course, that is partially because I am installing your program using python3 `pip` in a  `venv` environment. 

On hindsight, I wouldn't have installed Azure and Elevenlabs service had I known a little earlier that they don't work offline. Anyway, I can always uninstall them, can't I?!

Thank you and all the best, Mr. Beigel.

**Reply by KoljaB** [reply]
> Are there any offline TTS engine that could be used other than System TTS? System TTS is not very human-like.

Yes, CoquiEngine and ParlerEngine can be used offline. 

> The flipside is that the installation size appears huge, The download has exceeded 1GB filesize. 

The RealtimeTTS-master.zip has a download size of 58 MB, from which the biggest part are the two deepspeed wheels.

> On hindsight, I wouldn't have installed Azure and Elevenlabs service had I known a little earlier that they don't work offline. Anyway, I can always uninstall them, can't I?!

Sure, `pip uninstall RealtimeTTS` and they are gone. Azure and Elevenlabs are slim engines and don't take much space though, since the installed dependencies for both are not performing TTS for themselves and only call internet APIs.

**Reply by rajibando** [reply]
Apologise for overlooking to insert `Coqui TTS` though I had read what you had stated on your first post itself. Sorry!

>  The RealtimeTTS-master.zip has a download size of 58 MB, from which the biggest part are the two deepspeed wheels.

Yes, of course. But I didn't have pip and other dependencies installed.

Moreover, I wanted to install the program in a virtual environment `venv` not to leave my system in disarray. So the download size is over 3GB.

Still the installation got aborted because `pyaudio` wasn't installed. I am gradually installing the uninstalled components. An idea about the installation size is contained in the file:
[RealtimeTTSinPython3venv.txt](https://github.com/user-attachments/files/17893637/RealtimeTTSinPython3venv.txt)

>  Sure,  `pip uninstall RealtimeTTS` and they are gone. 

No, I didn't mean I would uninstall the  full program `RealtimeTTS`. Just the two said online TTS engines.

Thank you once more for replying to my posts.

**Reply by KoljaB** [reply]
> Yes, of course. But I didn't have pip and other dependencies installed.
> Moreover, I wanted to install the program in a virtual environment venv not to leave my system in disarray. So the download size is over 3GB.

I'm sorry, but I have to kindly disagree here. Whether you install RealtimeTTS in a virtual environment (venv) or not has no impact on the download size. However, it's true that RealtimeTTS includes some dependencies (libraries) depending on the installation variant, which can require additional disk space. These dependencies aren't part of my RealtimeTTS code directly, I have no control over the size of external libraries like Elevenlabs or Coqui TTS.

> Still the installation got aborted because pyaudio wasn't installed. I am gradually installing the uninstalled components. An idea about the installation size is contained in the file:

pyaudio is part of RealtimeTTS, but on Linux it requires portaudio to be installed first.

For Linux and CoquiEngine, I recommend:

```bash
# Update package lists
apt-get update

# Upgrade pip
python -m pip install --upgrade pip

# Install portaudio development package
apt-get install -y portaudio19-dev

# Install RealtimeTTS with Coqui
pip install RealtimeTTS[coqui]
```

> No, I didn't mean I would uninstall the full program RealtimeTTS. Just the two said online TTS engines.

I'm not certain if individual engines can be uninstalled separately. In this case, I would actually recommend running `pip uninstall RealtimeTTS` followed by `pip install RealtimeTTS[coqui]` to reinstall with just the engine you want.

**Reply by rajibando** [reply]
Thank you for your reply once more, Mr. Beigel,  I am really overwhelmed. Amidst your time crunch, the way you try to connect with your users and provide support is admirable.

> I'm sorry, but I have to kindly disagree here. Whether you install RealtimeTTS in a virtual environment (venv) or not has no impact on the download size.

Yes of course, you are obviously correct. You just looked at your project's installation size in isolation.
My way of looking was the packages I had to install for my system's python virtual environment for using your program. While you looked at the set, I looked at the super-set, the data download needed to set up the virtual environment to use your program.

So we really don't differ. Actually, only our perspectives are just different. Like those blindfolded men touching the elephant to determine what it is.

If the installation information is looked up in the text file attached earlier, it will clarify the position. Of course, your package doesn't take more space than what you say. As a creator you obviously know your creation the best.

> pyaudio is part of RealtimeTTS, but on Linux it requires portaudio to be installed first.

Yes, actually I shortened up the aspect by using a representative for the required add-ons.

My system Installed the following packages because of interconnected dependencies:

1. libasound2-dev (1.2.8-1+b1)
2. libjack-jackd2-dev (1.9.21~dfsg-3)
3. libpkgconf3 (1.8.1-1)
4. libportaudiocpp0 (19.6.0-1.2)
5. pkg-config (1.8.1-1)
6. pkgconf (1.8.1-1)
7. pkgconf-bin (1.8.1-1)
8. portaudio19-dev (19.6.0-1.2)

I usually keep a detailed record of the file size downloaded and the file size used up on my system. 

Thank you for your kind reply once again, Mr. Beigel,. Please take your time to reply back and spend your time upon your creation, until I report an emergency and request your kind attention for support.

> not certain if individual engines can be uninstalled separately. In this case, I would actually recommend running pip uninstall RealtimeTTS followed by pip install RealtimeTTS[coqui] to reinstall

Ok, understood your perspective. If those engines aren't large in size, I might then let them remain.

Thank you, have a good day, Mr. Beigel.


[Postscript]
Installed. I am skeptical of closed-source proprietary systems. Since the last time the installation process hadn't begun,  I used your clue and installed only `pip install RealtimeTTS[coqui]` this time.

I have attached the installation report with this post for you to peruse in case you want to.
[[KoljaB,RealtimeTTS]InstallReport.txt](https://github.com/user-attachments/files/17896794/KoljaB.RealtimeTTS.InstallReport.txt)

Ok, the installation is done. Now I have to figure the way to operate the TTS in the GUI environment.

Best wishes.
Rajib
